swift/ql/lib/codeql/swift/elements/expr/MethodLookupExprConstructor.qll

```diff
@@ -0,0 +1,17 @@
+private import codeql.swift.generated.Raw
+
+predicate constructMethodLookupExpr(Raw::SelfApplyExpr id) { any() }
+
+Raw::Decl extractDeclFromSelfApplyExpr(Raw::SelfApplyExpr e) {
+  exists(Raw::Expr unwrappedFunction | unwrappedFunction = unwrapConversion*(e.getFunction()) |
+    result =
+      [
+        unwrappedFunction.(Raw::DeclRefExpr).getDecl(),
+        unwrappedFunction.(Raw::OtherConstructorDeclRefExpr).getConstructorDecl()
+      ]
+  )
+}
+
+private Raw::Expr unwrapConversion(Raw::Expr e) {
+  e.(Raw::ImplicitConversionExpr).getSubExpr() = result
+}
```

Comment:
Does that work though? Would the type of `unwrapConversion*` then be `Raw::ImplicitConversionExpr → Raw::Expr`, or `Raw::Expr → Raw::Expr`? Because we want the latter, for the case when we pass a non-`Raw::ImplicitConversionExpr` argument, which we want to succeed.

Reply:
Yep, tried it, doesn't work. `unwrapConversion*(x)` is apparently not simply syntactic sugar for `[x, unwrapConversion+(x)]`.

Reply:
Oh, yes. You're right. Sorry for leading you down the wrong track. To see how they differ, I outputted the DIL for this:
```ql
Expr getLeft1(Expr e) {
  result = e.(MulExpr).getLeftOperand()
}

Expr getLeft2(MulExpr e) {
  result = e.getLeftOperand()
}

pragma[nomagic]
predicate test(Expr e, Expr left1, Expr left2) {
  left1 = getLeft1*(e) and
  left2 = getLeft2*(e)
}
```
and for the `left1 = getLeft1*(e)` conjunct the DIL looks like:
```ql
(left1 = e; getLeft#1Plus#ff(e, left1))
```
(which is DIL speak for `left1 = e or `left1 = getLeft+(e)`.)

And for the `left2 = getLeft2*(e)` conjunct we get:
```ql
(
  exists(cached int arg1 |
    arg1 = 27, exprs(e, arg1, _) // This checks that `e` is a `MulExpr`
  ),
  left2 = e
);
getLeft2#1Plus#ff(e, left2)
```
(which is DIL speak for `e instanceof MulExpr and left2 = e or getLeft2#1Plus#ff(e, left2`).

So you're right: `unwrapConversion*(x)` is not simply syntactic sugar for `[x, unwrapConversion+(x)]`. Rather, it's equivalent to `[x.(Raw::ImplicitConversionExpr), unwrapConversion+(x)]`. And indeed, what we want is your version.